src/monitors/LoopMonitor.v3

```diff
@@ -2,22 +2,52 @@
 // See LICENSE for details of Apache 2.0 license.
 
 // Implements a simple monitor that counts the number of iterations of each loop.
+class LmEntry(module: Module, func: FuncDecl, depth: int, pc: int) {
+	var endPc: int;
+}
+
 class LoopMonitor extends Monitor {
-	def entries = Vector<(Module, FuncDecl, int)>.new();
+	def entries = Vector<LmEntry>.new();
 	def counts = Vector<u64>.new();
 
 	def onParse(m: Module) {
 		var it = BytecodeIterator.new();
+
 		for (i < m.functions.length) {
 			var f = m.functions[i];
 			if (f.imp != null) continue; // skip imported functions
+
+			var loopDepth = 0;
+			var controlStack = ArrayStack<Opcode>.new();
+			var lmEntryStack = ArrayStack<LmEntry>.new();
+
 			for (bi = it.reset(f); bi.more(); bi.next()) {
-				if (bi.current() == Opcode.LOOP) {
-					var entry = entries.length;
-					m.insertProbeAt(f.func_index, bi.pc, LoopMonitorCounter.new(this, entry));
-					entries.put(m, f, bi.pc);
-					counts.put(0);
+				var op = bi.current();
+				match (op) {
+					BLOCK, IF => controlStack.push(op);
```

Comment:
Can you put in a TODO for `try` and exception-handling related bytecodes? Might be worthwhile to add a utility that automatically tracks the control stack in the future too.

Reply:
Added a TODO, will look at making the utility in the future (another PR)?